base-files: wifi: add random MAC support for wifi-iface

msvamp opened the pull request:
Add support for randomly generating a MAC address for a wifi-iface instance by setting `macaddr` to `random`
When set to `random`, a new locally administered unicast MAC address is generated and assigned to the iface everytime it is (re-)configured

This is typically useful for portable devices running OpenWrt as, using this option can mitigate attempts to track a device using its WiFi interface MAC address (especially for AP/STA interfaces), and ensure privacy of the user

mpratt14 commented:
see macaddr_random() in
package/base-files/files/lib/functions/system.sh

msvamp commented:
@mpratt14 Thanks!

So you suggest we include all the functions from `/lib/functions/system.sh` into `mac80211.sh`? Looks heavy.

msvamp commented:
Interestingly, I [don't see](https://github.com/search?q=org%3Aopenwrt+macaddr_random&type=code) `macaddr_random()` used anywhere in OpenWrt 😃

mpratt14 commented:
> Interestingly, I [don't see](https://github.com/search?q=org%3Aopenwrt+macaddr_random&type=code) `macaddr_random()` used anywhere in OpenWrt 😃

thats funny, not long ago it was being used

a8a17fd22326f0d1aa7644556eea6010e0779afe

functions like that are always kept, even if their original use case is not. you never know when downstream projects or a PR like yours needs it

mpratt14 commented:
import of the function script is done like so

`. /lib/functions/system.sh`

you can put that right before you call the function

also, we dont like backticks, please use the `$( ... )` syntax

msvamp commented:
> you can put that right before you call the function

Wouldn't that mean the functions will be re-defined every time a new random MAC address is being generated? (Say, a particular setup has multiple `wifi-iface`s with `macaddr random`

mpratt14 commented:
the difference should be negligible, but if you're worried about that, just put the import at the top of the script

mpratt14 commented:
before I forget, it should be noted that LuCI would not support the change here until at least this part is edited (there might be more changes in another file needed, not sure)

https://github.com/openwrt/luci/blob/master/modules/luci-base/htdocs/luci-static/resources/validation.js#L352

mpratt14 commented:
and by the way, don't close the PR, if you need help with Git just ask 👍🏼

msvamp commented:
> LuCI would not support the change here until at least this part is edited

Thanks for pointing this out, I was about to ask the same. I'll update LuCI next.

> and by the way, don't close the PR, if you need help with Git just ask 👍🏼

Definitely! No problem, I got this. I'll make the change and test it over the weekend.

mpratt14 commented on package/kernel/mac80211/files/lib/netifd/wireless/mac80211.sh:
```diff
@@ -667,10 +667,13 @@ mac80211_prepare_vif() {
 		macidx="$(($macidx + 1))"
-	}
+	elif [ "$macaddr" = 'random' ]; then
+		[ $(command -v 'macaddr_random') ] || source '/lib/functions/system.sh'
```
no need to check whether it is imported already, just do it

dot instead of source
`. /lib/functions/system.sh`

msvamp commented:
Ping 🙂

What is the recommended way to resolve the tiny conflict in Makefile? Should I update my base branch again and add the commit on top of it, and force push?

mpratt14 commented:
`git pull --rebase https://github.com/openwrt/openwrt.git master`

during the rebase, it will display that there is a conflict, more detail can be seen with `git status`

during a conflict, it shows both versions of the lines

between <<<< and ==== is the change from openwrt master

between ==== and >>>> is the change in your commit that is not in sync with openwrt master

good reading material
https://www.atlassian.com/git/tutorials/using-branches/merge-conflicts

msvamp committed:
base-files: wifi: add random MAC support for wifi-iface

Add support for randomly generating a MAC address for a wifi-iface
instance by setting `macaddr` to `random`
When set to `random`, a new locally administered unicast MAC address
is generated and assigned to the iface everytime it is (re-)configured

Signed-off-by: Manas Sambhus <manas.sambhus+github@gmail.com>

msvamp commented:
@mpratt14 I tried to rebase by accepting all current changes from `openwrt/master` except in the 2 files I changed, however it didn't seem to have worked out (please look at [this difference](https://github.com/openwrt/openwrt/compare/master...msvamp:openwrt:patch-2-failedrebase_09-08-22) between the resulting rebased branch and the current master).

I ended up re-creating the master of my fork from `openwrt/master`, re-adding the commit manually on top of it, and then force pushing to the source branch of this PR.

Neustradamus commented:
@msvamp: After merging in master, you can do a cherry-pick to 22.03 branch :)

aparcar commented:
To git.openwrt.org:openwrt/openwrt.git
   37c0d15a8e..1ea563ae13  master -> master

Please add the related change to LuCI.

msvamp commented:
> cherry-pick to 22.03 branch

@Neustradamus so a separate PR has to be created?

---

@aparcar Thank you for merging. I have already implemented the change in LuCI. It's here: https://github.com/openwrt/luci/pull/5896

aparcar commented:
> @Neustradamus so a separate PR has to be created?

No, just give it a few days to land on snapshots and proof to work, someone will do the backport
